sourcegraph/eclipse
# Webview Chat: basic implementation

Previously, there was no way to actually use the Chat inside Eclipse. This PR adds very basic Chat support to close this gap.

- When unauthenticated, there is a large "Login" button. After authenticating, the webview loads and you are able chat with Cody.
- When authenticated, the chat loads directly.

This PR completes several important technical milestones:

- We embed the Cody agent assets (index.js, *.wasm files, win-ca-roots.exe) with the plugin. We copy these assets from resources into the standard "data directory" of the "Cody Eclipse" program in the operating system.
- We embed the Cody webview assets (HTML, JS, CSS, SVG) with the plugin. We have a web server implementation that serves these assets directly from resources avoiding the need to copy them over to disk. This web servers does basic post-processing of index.html to inject missing CSS/JS to get the webview assets loading correctly, including setting up all the JVM/JS message passing infrastructre, and setting CSS styling variables so the UI looks similar to in VS Code.
- We launch the agent with the Node.js installation that is included with Eclipse. In the future, we probably need to embed Node.js with the plugin to support older Eclipse versions.
- The agent uses the server endpoint and access token that is provided in the onboarding flow. The access token is stored securely. After onboarding, the webview loads automatically (this took a surprisingly long time to solve).

The code is incredibly hacky and disorganized. The scope of this PR is just to get the basic functionality working to validate the technical feasibility of this approach. We have a lot of work left to actually clean up the implementation.


## Test plan

* Build the plugin locally
* Click Login
* Click "OK"
* Make sure that the webview loads after authorizing
* Send chat message, confirm you get a reply
![image](https://github.com/sourcegraph/eclipse/assets/1408093/2db78317-47b6-4061-909c-494ae385bc1f)
